glommio/src/read_write_lock.rs

```diff
@@ -0,0 +1,1214 @@
+        let mut state = self.rw.borrow_mut();
+        match self.kind {
+            WaiterKind::WRITER => {
+                if state.try_write()? {
+                    Poll::Ready(Ok(()))
+                } else {
+                    state.add_waker(self.id, self.kind, cx.waker().clone());
+                    Poll::Pending
+                }
+            }
+
+            WaiterKind::READER => {
+                if state.try_read()? {
+                    Poll::Ready(Ok(()))
+                } else {
+                    state.add_waker(self.id, self.kind, cx.waker().clone());
+                    Poll::Pending
+                }
+            }
+        }
+    }
+}
+
+impl<T> State<T> {
+    fn new(value: T) -> Self {
+        State {
+            value: UnsafeCell::new(value),
+            id_gen: 0,
+            writers: 0,
+            readers: 0,
+            queued_writers: 0,
+            closed: false,
+
+            waiters: VecDeque::new(),
+            waiters_map: HashMap::new(),
+        }
+    }
+
+    fn try_read(&mut self) -> ReadWriteLockResult<bool> {
+        if self.closed {
+            return Err(LockClosedError);
+        }
+
+        debug_assert!(!(self.readers > 0 && self.writers > 0));
+
+        if self.writers == 0 {
+            self.readers += 1;
+            return Ok(true);
+        }
+
+        Ok(false)
+    }
+
+    fn try_write(&mut self) -> ReadWriteLockResult<bool> {
+        if self.closed {
+            return Err(LockClosedError);
+        }
+
+        debug_assert!(!(self.readers > 0 && self.writers > 0));
+
+        if self.readers == 0 {
+            self.writers += 1;
+            return Ok(true);
+        }
+
+        Ok(false)
+    }
+
+    fn add_waker(&mut self, id: WaiterId, kind: WaiterKind, waker: Waker) {
+        debug_assert!(!(self.readers > 0 && self.writers > 0));
+
+        let entry = self.waiters_map.entry(id);
+        if let Vacant(entry) = entry {
+            entry.insert((kind, waker));
+            self.waiters.push_back(id);
+
+            if kind == WaiterKind::WRITER {
+                self.queued_writers += 1;
+            }
+        }
+    }
+}
+
+/// RAII structure used to release the shared read access of a lock when
+/// dropped.
+///
+/// This structure is created by the [`read`] and [`try_read`] methods on
+/// [`ReadWriteLock`].
+///
+/// [`read`]: ReadWriteLock::read
+/// [`try_read`]: ReadWriteLock::try_read
+#[derive(Debug)]
+#[must_use = "if unused the ReadWriteLock will immediately unlock"]
+pub struct ReadWriteLockReadGuard<'a, T> {
+    rw: Rc<RefCell<State<T>>>,
+    //guard can not outlive rw-lock owner
+    phantom: PhantomData<&'a T>,
+}
+
+impl<'a, T> Deref for ReadWriteLockReadGuard<'a, T> {
+    type Target = T;
+
+    fn deref(&self) -> &Self::Target {
+        let state = (*self.rw).borrow();
+        if state.closed {
+            panic!("Related ReadWriteLock is already closed");
+        }
+
+        let ptr = state.value.get();
```

Comment:
do we really need to use an `UnsafeCell` here? I am wondering if we couldn't use a `RefCell` and move the `RefMut`inside the `ReadWriteLockxxGuard`.

I don't feel too strongly, but @matklad has been cracking down hard on our `UnsafeCell`!

Reply:
Not taking into account that is a small but unnecessary slowdown, unsafe API is very localized, I had a feeling that I will not be able to rid off Ref/RefMut and they will be exposed to the user API, which contradicts with the meaning of "clear"  API. That was the feeling which I have got from my previous experience. Those things quity sticky once you start to work with them as in this example https://rust-unofficial.github.io/too-many-lists/fourth-peek.html .  But I will try to fix that. Maybe that is very easy and my feelings are wrong.

Reply:
Hi @glommer @matklad , I tried to use RefCell both as a wrapper on the value and using top-level RefCell in both cases I get  `returns a value referencing data owned by the current function` the same as in this example https://rust-unofficial.github.io/too-many-lists/fourth-peek.html also I tend to agree with the author of example that in such case Ref/RefMut can not be hidden from the user. I prefer to do not expose internals to the user and add one more level of indirection. UnsafeCell in this context is used in absolutely the same way as it is done in RwLock, unsafe code is quite localized and as result not dangerous. If you do not mind I prefer to keep UnsafeCell. Looking for your feedback :-)

Reply:
Hi, @glommer @matklad all problems with UnsafeCell are boiled down to the single one - `get_mut` can not be implemented without exposing of Ref/RefMut to the client, the reason is quite simple in such case reference which is owned by Ref/RefMut is exposed to the client while Ref is dropped in the function. If you think that is OK to expose Ref or better IMHO WriteGuard to the client I will do that otherwise I think UnsafeCell should be kept in the code.

Reply:
>get_mut can not be implemented without exposing of Ref/RefMut to the client,

I mean, we can do this with `unsafe`, but without `UnsafeCell`: `unsafe { &*(&state.value as *const _) }`. We fundamentally need unsafe *somewhere*, as we are implementing interior mutability primitive. The question is about minimizing it, not eliminating (in this case). 

Although I am not entirely clear on safety implications of either approach, I'll take a thorough look once I have resources for a thorough look.

One thing that gives me a pause though, is that std's MutexGuard looks like this:

```rust
pub struct MutexGuard<'a, T: ?Sized + 'a> {
    lock: &'a Mutex<T>,
    poison: poison::Guard,
}
```

Naively, I would expect to see a similar lifetime setup here: a reference instead of a phantom data. Intuitively, I don't think there should be differences in how ownership is setup between async and sync cases. But this is intuition, I might be wrong.

Reply:
Hi @matklad . I have noticed this approach too. I thought that in our case a reference to the RwLock is at the fact the reference to its state but without additional indirection + PhantomData would cost nothing in that case. That is why I used the state field directly.  As about the get_mut I understood your idea. As for guards, I plan to incorporate Ref/RefMut as a field which will remove the need to have PhantomData and unsafe code at all. But in such case, I suppose there is no need to use a reference to the lock itself as the lifetime will be dictated by the Ref lifetime. Do you have any objections?

Reply:
I mean if I would use Ref to the state, actually I suppose I need even RefMut, acquisitions of reading accesses to the lock would panic during switching of execution inside of fibres.